File: constraints/src/main/java/gov/nasa/jpl/aerie/constraints/model/LinearEquation.java
package gov.nasa.jpl.aerie.constraints.model;

import gov.nasa.jpl.aerie.constraints.time.Interval;
import gov.nasa.jpl.aerie.constraints.time.Segment;
import gov.nasa.jpl.aerie.constraints.time.Windows;
import gov.nasa.jpl.aerie.merlin.protocol.types.Duration;

import java.util.Objects;
import java.util.Optional;
import java.util.function.BiFunction;

import static gov.nasa.jpl.aerie.constraints.time.Interval.Inclusivity.Exclusive;
import static gov.nasa.jpl.aerie.constraints.time.Interval.Inclusivity.Inclusive;

/**
 * A linear equation in point-slope form.
 */
public final class LinearEquation {
  public final Duration initialTime;
  public final double initialValue;
  public final double rate;

  public LinearEquation(final Duration initialTime, final double initialValue, final double rate) {
    this.initialTime = initialTime;
    this.initialValue = initialValue;
    this.rate = rate;
  }

  public double valueAt(final Duration time) {
    return this.initialValue + this.rate*(time.minus(this.initialTime)).ratioOver(Duration.SECOND);
  }

  public LinearEquation shiftInitialTime(final Duration newInitialTime) {
    return new LinearEquation(
        newInitialTime,
        this.initialValue + newInitialTime.minus(this.initialTime).ratioOver(Duration.SECOND)*this.rate,
        this.rate
    );
  }

  public boolean changing() {
    return this.rate != 0;
  }

  private Optional<Duration> intersectionPointWith(final LinearEquation other) {
    if (this.rate == other.rate) return Optional.empty();

    /*
    Floating point noise can cause rates to be extremely near zero, when they should
    have been exactly zero. For example: `0.1 + 0.2 - 0.1 - 0.2 != 0`.

    This can cause the denominator below to be tiny, leading to long overflow later.
     */

    // If the following causes an exception, something really has gone wrong, and we don't want to catch it.
    final double numSeconds = (other.valueAt(this.initialTime) - this.initialValue) / (this.rate - other.rate);

    // Check if numSeconds is too big before putting it in a long.
    if (Math.abs(numSeconds) > ((double) Long.MAX_VALUE) / Duration.SECOND.dividedBy(Duration.MICROSECOND)) {
      return Optional.empty();
    }

    return Optional.of(
        this.initialTime.plus(
            Duration.roundNearest(
                numSeconds,
                Duration.SECONDS
            )
        )
    );
  }

  public Windows intervalsLessThan(final LinearEquation other) {
    return getInequalityIntervals(other, (l, r) -> l < r);
  }

  public Windows intervalsLessThanOrEqualTo(final LinearEquation other) {
    return getInequalityIntervals(other, (l, r) -> l <= r);
  }

  public Windows intervalsGreaterThan(final LinearEquation other) {
    return getInequalityIntervals(other, (l, r) -> l > r);
  }

  public Windows intervalsGreaterThanOrEqualTo(final LinearEquation other) {
    return getInequalityIntervals(other, (l, r) -> l >= r);
  }

  private Windows getInequalityIntervals(
      final LinearEquation other,
      final BiFunction<Double, Double, Boolean> op
  ) {
    final var intersectionOption = this.intersectionPointWith(other);

    if (intersectionOption.isEmpty()) {
      final var resultEverywhere = op.apply(this.initialValue, other.valueAt(this.initialTime));
      return new Windows(resultEverywhere);
    } else {
      final var intersection = intersectionOption.get();
      final var oneSecondBefore = intersection.minus(Duration.SECOND);
      final var oneSecondAfter = intersection.plus(Duration.SECOND);
      return new Windows(
          Segment.of(
              Interval.between(Duration.MIN_VALUE, Inclusive, intersection, Exclusive),
              op.apply(this.valueAt(oneSecondBefore), other.valueAt(oneSecondBefore))
          ),
          Segment.of(
              Interval.at(intersection),
              op.apply(this.valueAt(intersection), other.valueAt(intersection))
          ),
          Segment.of(
              Interval.between(intersection, Exclusive, Duration.MAX_VALUE, Inclusive),
              op.apply(this.valueAt(oneSecondAfter), other.valueAt(oneSecondAfter))
          )
      );
    }
  }

  public Windows intervalsEqualTo(final LinearEquation other) {
    final var intersection = this.intersectionPointWith(other);

    if (intersection.isEmpty()) {
      return new Windows(this.initialValue == other.valueAt(this.initialTime));
    } else {
      final var t = intersection.get();
      return new Windows(
          Segment.of(Interval.between(Duration.MIN_VALUE, Inclusive, t, Exclusive), false),
          Segment.of(Interval.at(t), true),
          Segment.of(Interval.between(t, Exclusive, Duration.MAX_VALUE, Inclusive), false)
      );
    }
  }

  public Windows intervalsNotEqualTo(final LinearEquation other) {
    return this.intervalsEqualTo(other).not();
  }

  public static Windows lessThan(LinearEquation left, LinearEquation right) {
    return left.intervalsLessThan(right);
  }

  public static Windows lessThanOrEqualTo(final LinearEquation left, final LinearEquation right) {
    return left.intervalsLessThanOrEqualTo(right);
  }

  public static Windows greaterThan(final LinearEquation left, final LinearEquation right) {
    return left.intervalsGreaterThan(right);
  }

  public static Windows greaterThanOrEqualTo(final LinearEquation left, final LinearEquation right) {
    return left.intervalsGreaterThanOrEqualTo(right);
  }

  public static Windows equalTo(final LinearEquation left, final LinearEquation right) {
    return left.intervalsEqualTo(right);
  }

  public static Windows notEqualTo(final LinearEquation left, final LinearEquation right) {
    return left.intervalsNotEqualTo(right);
  }

  public String toString() {
    return String.format(
            "{\n" +
            "  Initial Time: %s\n" +
            "  Initial Value: %s\n" +
            "  Rate: %s\n" +
            "}",
            this.initialTime,
            this.initialValue,
            this.rate
    );
  }

  @Override
  public boolean equals(final Object obj) {
    if(!(obj instanceof final LinearEquation other)) return false;

    return this.valueAt(Duration.ZERO) == other.valueAt(Duration.ZERO)
        && this.valueAt(Duration.MINUTE) == other.valueAt(Duration.MINUTE);
  }

  @Override
  public int hashCode() {
    return Objects.hash(this.initialValue, this.initialTime, this.rate);
  }
}

File: procedural/timeline/src/main/kotlin/gov/nasa/ammos/aerie/procedural/timeline/payloads/LinearEquation.kt
package gov.nasa.ammos.aerie.procedural.timeline.payloads

import gov.nasa.jpl.aerie.merlin.protocol.types.Duration
import gov.nasa.ammos.aerie.procedural.timeline.Interval
import gov.nasa.ammos.aerie.procedural.timeline.Interval.Inclusivity.Exclusive
import gov.nasa.ammos.aerie.procedural.timeline.Interval.Inclusivity.Inclusive
import gov.nasa.ammos.aerie.procedural.timeline.collections.profiles.Real
import gov.nasa.ammos.aerie.procedural.timeline.collections.profiles.Booleans
import gov.nasa.ammos.aerie.procedural.timeline.util.duration.div
import java.util.function.BiFunction
import kotlin.math.abs
import kotlin.math.absoluteValue

/** A linear equation in point-slope form. */
data class LinearEquation(
    /** The time of the start point. */
    @JvmField val initialTime: Duration,
    /** The value of the start point. */
    @JvmField val initialValue: Double,
    /** The rate of change, in units per second. */
    @JvmField val rate: Double
) {
  /** Creates a constant linear equation at a given value. */
  constructor(constant: Number): this(Duration.ZERO, constant.toDouble(), 0.0)

  /** Calculates the value at a given time. */
  fun valueAt(time: Duration): Double {
    val change = rate * time.ratioOver(Duration.SECOND) - rate * initialTime.ratioOver(Duration.SECOND)
    return initialValue + change
  }

  /** Returns an equivalent equation that is represented at a different start time. */
  fun shiftInitialTime(newInitialTime: Duration) = LinearEquation(
      newInitialTime,
      initialValue + newInitialTime.minus(initialTime).ratioOver(Duration.SECOND) * rate,
      rate
  )

  /***/
  fun isConstant() = rate == 0.0

  private fun intersectionPointWith(other: LinearEquation): Duration? {
    if (rate == other.rate) return null

    /*
    Floating point noise can cause rates to be extremely near zero, when they should
    have been exactly zero. For example: `0.1 + 0.2 - 0.1 - 0.2 != 0`.

    This can cause the denominator below to be tiny, leading to long overflow later.
     */

    // If the following causes an exception, something really has gone wrong, and we don't want to catch it.
    val numSeconds = (other.valueAt(initialTime) - initialValue) / (rate - other.rate)

    // Check if numSeconds is too big before putting it in a long.
    return if (abs(numSeconds) > Long.MAX_VALUE.toDouble() / (Duration.SECOND / Duration.MICROSECOND)) null
    else initialTime.plus(Duration.roundNearest(numSeconds, Duration.SECOND))
  }

  /** Calculates when this is less than another linear equation, as a [Booleans] object. */
  infix fun intervalsLessThan(other: LinearEquation): Booleans {
    return getInequalityIntervals(other) { l: Double, r: Double -> l < r }
  }

  /** Calculates when this is less than or equal to another linear equation, as a [Booleans] object. */
  infix fun intervalsLessThanOrEqualTo(other: LinearEquation): Booleans {
    return getInequalityIntervals(other) { l: Double, r: Double -> l <= r }
  }

  /** Calculates when this is greater than another linear equation, as a [Booleans] object. */
  infix fun intervalsGreaterThan(other: LinearEquation): Booleans {
    return getInequalityIntervals(other) { l: Double, r: Double -> l > r }
  }

  /** Calculates when this is greater than or equal to another linear equation, as a [Booleans] object. */
  infix fun intervalsGreaterThanOrEqualTo(other: LinearEquation): Booleans {
    return getInequalityIntervals(other) { l: Double, r: Double -> l >= r }
  }

  private fun getInequalityIntervals(
      other: LinearEquation,
      op: BiFunction<Double, Double, Boolean>
  ): Booleans {
    val intersection = intersectionPointWith(other)
    return if (intersection === null) {
      val resultEverywhere = op.apply(initialValue, other.valueAt(initialTime))
      Booleans(resultEverywhere)
    } else {
      val oneSecondBefore = intersection.minus(Duration.SECOND)
      val oneSecondAfter = intersection.plus(Duration.SECOND)
      Booleans(
          Segment(
              Interval.betweenClosedOpen(Duration.MIN_VALUE, intersection),
              op.apply(this.valueAt(oneSecondBefore), other.valueAt(oneSecondBefore))
          ),
          Segment(
              Interval.at(intersection),
              op.apply(this.valueAt(intersection), other.valueAt(intersection))
          ),
          Segment(
              Interval.between(intersection, Duration.MAX_VALUE, Exclusive, Inclusive),
              op.apply(this.valueAt(oneSecondAfter), other.valueAt(oneSecondAfter))
          )
      )
    }
  }

  /** Calculates when this is equal to another linear equation, as a [Booleans] object. */
  fun intervalsEqualTo(other: LinearEquation): Booleans {
    val intersection = intersectionPointWith(other)
    return if (intersection === null) {
      Booleans(initialValue == other.valueAt(initialTime))
    } else {
      Booleans(
          Segment(Interval.betweenClosedOpen(Duration.MIN_VALUE, intersection), false),
          Segment(Interval.at(intersection), true),
          Segment(Interval.between(intersection, Duration.MAX_VALUE, Exclusive, Inclusive), false)
      )
    }
  }

  /** Calculates when this is not equal to another linear equation, as a [Booleans] object. */
  fun intervalsNotEqualTo(other: LinearEquation): Booleans {
    return !intervalsEqualTo(other)
  }

  /** Finds the time that this equation is zero, or `null` if it does not cross the axis. */
  fun findRoot() = if (rate == 0.0) null else initialTime - Duration.roundNearest(initialValue / rate, Duration.SECOND)

  /** Calculates the absolute value of this equation, as a real profile. */
  fun abs() =
      if (isConstant()) Real(LinearEquation(initialTime, initialValue.absoluteValue, 0.0))
      else {
        val root = findRoot()!!
        Real(
            Segment(Interval.betweenClosedOpen(Duration.MIN_VALUE, root), LinearEquation(root, 0.0, -rate.absoluteValue)),
            Segment(Interval.between(root, Duration.MAX_VALUE), LinearEquation(root, 0.0, rate.absoluteValue))
        )
      }

  /***/
  override fun toString(): String {
    return String.format(
        """{
  Initial Time: %s
  Initial Value: %s
  Rate: %s
}""",
        initialTime,
        initialValue,
        rate
    )
  }

  /***/
  override fun equals(other: Any?) =
      if (other !is LinearEquation) false
      else initialValue == other.valueAt(initialTime) && rate == other.rate

  /***/
  override fun hashCode(): Int {
    var result = initialTime.hashCode()
    result = 31 * result + initialValue.hashCode()
    result = 31 * result + rate.hashCode()
    return result
  }
}
